### ISSUE TITLE
Possible timing attack vulnerability

### ISSUE THREAD

**Issue by Philipp91** [loggerhead/shadowsocks-rust#2]
Just a quick heads-up from a passer-by:
https://github.com/loggerhead/shadowsocks-rust/blob/master/src/crypto/encryptor.rs#L314
This comparison might be vulnerable to timing attacks because it's not constant-time. Not sure if that's relevant here at all. I'm reporting such issues to a couple of repositories, please just ignore/close if it isn't applicable here.

**Comment by loggerhead** [comment]
@Philipp91 Sorry, I am not familiar with security things. You mean because the length of `data` or `key` is not a fix size, so the compute of hmac is not a constant-time, which will make the application be vulnerable to timing attacks?

**Comment by Philipp91** [comment]
I apologize for being so unspecific. So the `.eq()` call compares two `&[u8]` slices. Let's assume that they have the same length (otherwise they can't be equal anyway). The comparison runs byte-by-byte and aborts when it finds two non-equal bytes. After checking all bytes successfully, it would return true. So the longer the common prefix of the slices, the longer the runtime; and if they're equal, the runtime is longest. Attackers can, in some scenarios, use this for timing attacks and then guess each byte individually, because they know that they've got the prefix before that right already.

Again, I have no idea if such an attack would work in the context that you use this `.eq()` call in. But in general, it is best practice to use constant-time comparisons for digests.

You could change the return value of `hmac_sha()` to `MacResult` and use `result()` instead of `raw_result()`. To compare, you convert your other slice to `MacResult::new(sha1)` and then you simply use `==` or `.eq()` on the `MacResult` instances. This ends up using this constant-time comparison here:
https://github.com/DaGenix/rust-crypto/blob/master/src/mac.rs#L88

Alternatively, you could call the `crypto::util::fixed_time_eq()` function directly with your slices.

**Comment by loggerhead** [comment]
Thank you for your explanation and suggestions, I will be careful with this problem on my future projects 😃 , but I am not gone to fix this problem in shadowsocks-rust. 

For shadowsocks-rust, the equal comparison of hmac only happens when one time auth enabled, which means the hmac of each packet is **unique**, so this kind of attack should not success on shadowsocks-rust. On the other hand, shadowsocks-rust care about whether your messages can secret arrive the receiver, but not promise the security of your messages (which should be the responsibility of application).

**Comment by quininer** [comment]
> which means the hmac of each packet is unique, so this kind of attack should not success on shadowsocks-rust

No, attacker could use this vulnerability to construct any packet and pass one time auth (just MAC). while this is practically impossible in reality, avoiding it is common sense.
You can use [this challenge](http://cryptopals.com/sets/4/challenges/31) to understand the "insecure compare" risk.

**Comment by loggerhead** [comment]
@quininer 
> No, attacker could use this vulnerability to construct any packet and pass one time auth (just MAC).

 Sorry, I understand the risk of the "insecure compare", but I don't understand how to implementation the attack you described above. As the attacker don't know the key (need `password` to construct) used in encryption, so he can't construct a correct packet with a custom hmac that `ssserver` can correctly decrypt, so he can't construct "insecure compare" as he wanted.

**Comment by quininer** [comment]
@loggerhead see https://github.com/OWASP/railsgoat/wiki/A2-Insecure-Compare-and-Timing-Attacks

**Comment by quininer** [comment]
I reconsidered the problem, MtE **can't defend** it.
now, let's speak chinese.


Alice 和 Carol 之間使用該協議進行通信，但 Carol 使用了非常數時間比較 MAC。
假設有一敵手 Mallory，他能竊聽雙方通信，並分析構造和發送數據包。

MtE 不能防禦非常數時間比較導致的時序攻擊
-----

```
+----------+-----------+----------+
| DATA.LEN | HMAC-SHA1 |   DATA   |
+----------+-----------+----------+
|     2    |     10    | Variable |
+----------+-----------+----------+
```

當 Alice 和 Carol 使用流密碼時，Mallory 修改密文其中一個 byte 不會影響到其他數據。
寫作 `E(H) + E(M') == E(H + M')`。

1. Mallory 截獲一條密文 `E(H + M)`，並修改 DATA 部分構造 `E(H + M')`。
2. Carol 將其解密，獲得 `H + M'`，顯然 `H != HMAC(M')`，
        但由於使用了 "insecure compare"， 泄漏了 `H` 與 `HMAC(M')` 的關係信息。
3. Mallory 可以修改 HMAC-SHA1 部分構造 `E(H' + M')`，並發送至 Carol。
        Mallroy 不需要知道 `H'` 的值，只需要逐字節修改密文並判斷出耗時最長的 byte。
4. 重複 3，令 `H'` 逼近 `HMAC(M')`，直到 `H' == HMAC(M')`。

**Comment by loggerhead** [comment]
@quininer Thanks for your patience, you are right, this attack is possible. I will fix this issue in next version (0.7.0), thank you again 👍. 

But I want clarify that security is __NOT__ the first citizen in shadowsocks. Shadowsocks is used to bypass the high wall, which means the connection or `ssserver` without obvious features is __MUCH__ important, and that's the first citizen in shadowsocks 😃.

**Comment by loggerhead** [comment]
Thanks, the problem is fixed in v0.6.2.
